docs/applications/nublado/major-upgrade.rst

```diff
@@ -0,0 +1,313 @@
+
+Upgrading nublado v2 to nublado v3 is a complex process, and is usually
+done in stages; in the general case, each of these three stages would
+take place in a separate maintenance window as defined by site policy:
+
+#. Enable nublado v3 on a non-default route (i.e. not ``/nb``; in this
+   document and in actuality, we use ``/n3``).  Run for some time
+   allowing users to test the new spawner while still using nublado v2
+   by default
+#. Switch nublado v3 to the default route and nublado v2 to an alternate
+   route (e.g. ``/n2``).  Again, run for some time, keeping an eye out
+   for any problems.
+#. Once satisfied with the performance of nublado v3, disable the
+   ``nublado2``, ``cachemachine``, and ``moneypenny`` applications (the
+   functionality of the latter two are subsumed in nublado v3).
+
+Enable nublado v3
+-----------------
+
+Pre-deployment Preparation
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The following tasks can be performed at any time, and should be
+performed ahead of your maintenance window to actually turn on the
+application.
+
+Vault Secret Creation
+"""""""""""""""""""""
+
+The first thing you will need to do is to set up secrets for
+``nublado`` (in future versions, this will require copying ``nublado``
+to ``nubladoN``, but ``nublado2`` is the current name for version 2, so
+if you're running an RSP, you already have those secrets set up
+correctly).
+
+These secrets, stored in Vault, and presented by
+``vault-secrets-operator`` are:
+
+#. ``cryptkeeper_key``: may be randomly generated; used internally by
+   the Hub to encrypt authentication state in the session database.
+#. ``crypto_key``: may be randomly generated; used to encrypt cookies
+   for user state in the browser session.
+#. ``proxy_token``: may be randomly generated; used to encrypt
+   communications between the JupyterHub Configurable HTTP Proxy and the
+   Hub.
+#. ``hub_db_password``: the password for the Hub session database; in
+   nublado v3 this is a Postgres database; conventionally both the
+   username and the database name are ``nublado3``.  In general you
+   should be consuming a database as a service, in which case you will
+   need to request this database from your DBAs and ask them the
+   password.  If you are instead running the internal database, you will
+   need to create this database and grant permissions in the
+   ``postgres`` application configuration.
+#. ``slack_webhook``: the Slack webhook allowing the JupyterLab
+   controller to post status messages to Slack.  This is the same one as
+   used by, e.g. Gafaelfawr in your RSP instance.
+
+Note that in future the secret setup and management will all be handled
+as part of Phalanx deployment, and you will no longer need to do this
+manually.  However, at the present, you still do.
+
+Enable the application
+""""""""""""""""""""""
+
+In the values file for your installation in the top-level Phalanx
+``environments`` directory, set ``nublado: true``.
```

Comment:
This will probably always and already be set? Are you going to copy over the whole /applications/nublado around in phalanx?

This seems like something else that the user doesn't see, and moving these directories around seems like it could be really hard to back out a change if there's a problem?

Reply:
There are going to be Phalanx environments without nublado.  Roundtable, for instance.  Encoding the major version in the app name for future releases is probably a good idea, although I suspect we had some theory for why we shouldn't this time that I have forgotten.